### ISSUE TITLE
Don’t list redis as stable backend and/or link to issues/caveats

### ISSUE THREAD

**Issue by torotil** [celery/celery#8568]
# Checklist

- [x] I have checked the [issues list](https://github.com/celery/celery/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3A%22Category%3A+Documentation%22+)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/main)
  to find out if the bug was already fixed in the main branch.
- [x] I have included all related issues and possible duplicate issues in this issue
      (If there are none, check this box anyway).

#### Related Issues

- #4766 already added some warnings regarding the eta/countdown feature and redis.
- #1810 and the related https://github.com/celery/kombu/issues/533 is a reproducible bug in kombu that leads to data-loss that is unsolved since 2015.

#### Possible Duplicates

- None

# Description

The [broker overview](https://docs.celeryq.dev/en/stable/getting-started/backends-and-brokers/index.html) lists Redis as one of the two stable backends that support all features.

This is although there are major caveats when using Redis especially when using ETA tasks. At least one of them also leads to data loss. Some of those caveats are listed in [the Caveats section on the “Using Redis” page](https://docs.celeryq.dev/en/stable/getting-started/backends-and-brokers/redis.html#caveats) others not. None of them are hinted at on the broker overview page.

To be fair there is one sentence on the [“First steps with celery”](https://docs.celeryq.dev/en/latest/getting-started/first-steps-with-celery.html#redis) page that says:

> [Redis](https://redis.io/) is also feature-complete, but is more susceptible to data loss in the event of abrupt termination or power failures.

This is very vague and it is not linked to in the broker overview page.

# Suggestions

The broker overview page should not list Redis as a stable backend or at least should directly link to a page detailing all the caveats and long-term issues with Redis. It should also explain their consequences: Which features you can’t use or can only use with major restrictions.

**Comment by auvipy** [comment]
I agree with you. can you help to change the doc about redis? In recent times I found other back ends more stable then redis

**Comment by Nusnus** [comment]
Unfortunately I also agree.
On the bright side, the upcoming pytest-celery plugin v1.0.0 allows (finally) to automatically test broker reconnection (also with Redis) so we'll be a huge step forward into making Redis indeed feature complete.

When the time comes, I might take the challenge and fix it myself, so if you can help as @auvipy said with maybe collecting the top issues to add to the doc I'll be able to use that as a todo-list to get to feature complete status (again?).

**Comment by auvipy** [comment]
I can collaborate to reach that target as well. thanks for heads up Nusnus

**Comment by Nusnus** [comment]
> I can collaborate to reach that target as well. thanks for heads up Nusnus

🚀🚀🚀

**Comment by Nusnus** [comment]
@torotil @auvipy 
> ## Related Issues and Possible Duplicates
> #### Related Issues
> * [docs regarding how to use eta/countdown should mention visibility_timeout setting for Redis and SQS brokers #4766](https://github.com/celery/celery/issues/4766) already added some warnings regarding the eta/countdown feature and redis.
> * [ETA tasks lost with redis #1810](https://github.com/celery/celery/issues/1810) and the related [ETA tasks lost on worker TERM restart kombu#533](https://github.com/celery/kombu/issues/533) is a reproducible bug in kombu that leads to data-loss that is unsolved since 2015.
>
> ...
>
> # Description
> The [broker overview](https://docs.celeryq.dev/en/stable/getting-started/backends-and-brokers/index.html) lists Redis as one of the two stable backends that support all features.
> 
> This is although there are major caveats when using Redis especially when using ETA tasks. At least one of them also leads to data loss. Some of those caveats are listed in [the Caveats section on the “Using Redis” page](https://docs.celeryq.dev/en/stable/getting-started/backends-and-brokers/redis.html#caveats) others not. None of them are hinted at on the broker overview page.
> 
> To be fair there is one sentence on the [“First steps with celery”](https://docs.celeryq.dev/en/latest/getting-started/first-steps-with-celery.html#redis) page that says:
> 
> > [Redis](https://redis.io/) is also feature-complete, but is more susceptible to data loss in the event of abrupt termination or power failures.
> 
> This is very vague and it is not linked to in the broker overview page.
>
> Looking around there are quite a few issues on this issue queue and questions on stackoverflow that come to the conclusion that switching to RabbitMQ solves the issue — or debugging the Redis backend. This leads me to the conclusion that the Redis backend is not in fact stable.
> 
> # Suggestions
> The broker overview page should not list Redis as a stable backend or at least should directly link to a page detailing all the caveats and long-term issues with Redis. It should also explain their consequences: Which features you can’t use or can only use with major restrictions.

Celery v5.5, currently in [beta](https://github.com/celery/celery/releases/tag/v5.5.0b2), will maintain much higher-quality support with Redis, finally handling long-due issues.
1. [Kombu v5.4.0](https://github.com/celery/kombu/releases/tag/v5.4.0) finally fixes the reconnection bug with Redis, where the worker stopped consuming messages. This bug existed for a few years.
2. The new [Soft Shutdown](https://docs.celeryq.dev/en/latest/userguide/workers.html#worker-shutdown) and [new config](https://docs.celeryq.dev/en/latest/userguide/configuration.html#worker-soft-shutdown-timeout) finally allow mitigation of lost tasks, [including ETA tasks](https://github.com/celery/celery/pull/9231), during worker teardown. This addresses the [linked issues from 2015](https://github.com/celery/celery/issues/1810). These are also relevant for SQS btw.
3. A new caveat was added in both [Redis](https://docs.celeryq.dev/en/latest/getting-started/backends-and-brokers/redis.html#soft-shutdown) and SQS docs regarding the usefulness of the soft shutdown regarding task re-queue.
4. Many new smoke tests were added to verify that Redis's behavior matches RabbitMQ regarding the above points, so it is also tested via the automatic CI.

We can close this issue when Celery v5.5.0 (using Kombu v5.4.0) is released.
This doesn’t mean Redis support is perfect, but we can declare it “stable” for the time being.

**Comment by Nusnus** [comment]
Closing for now, we may reopen if anyone disagrees that the issue should be closed.
